rls/src/main/java/io/grpc/rls/internal/CachingRlsLbClient.java

```diff
@@ -0,0 +1,970 @@
+      RouteLookupRequest request =
+          requestFactory.create(methodName[0], methodName[1], args.getHeaders());
+      final CachedRouteLookupResponse response = CachingRlsLbClient.this.get(request);
+
+      PickSubchannelArgs rlsAppliedArgs = getApplyRlsHeader(args, response);
+      if (response.hasData()) {
+        ChildPolicyWrapper childPolicyWrapper = response.getChildPolicyWrapper();
+        ConnectivityState connectivityState =
+            childPolicyWrapper.getConnectivityStateInfo().getState();
+        switch (connectivityState) {
+          case CONNECTING:
+            return PickResult.withNoResult();
+          case IDLE:
+            // fall through
+          case READY:
+            if (childPolicyWrapper.getPicker() == null) {
+              return PickResult.withNoResult();
+            }
+            return childPolicyWrapper.getPicker().pickSubchannel(rlsAppliedArgs);
+          case TRANSIENT_FAILURE:
+            return handleError(rlsAppliedArgs, Status.INTERNAL);
+          case SHUTDOWN:
+          default:
+            return handleError(rlsAppliedArgs, Status.ABORTED);
+        }
+      } else if (response.hasError()) {
+        return handleError(rlsAppliedArgs, response.getStatus());
+      } else {
+        return PickResult.withNoResult();
+      }
+    }
+
+    private PickSubchannelArgs getApplyRlsHeader(
+        PickSubchannelArgs args, CachedRouteLookupResponse response) {
+      if (response.getHeaderData() == null || response.getHeaderData().isEmpty()) {
+        return args;
+      }
+
+      Metadata headers = new Metadata();
+      headers.merge(args.getHeaders());
+      headers.put(RLS_DATA_KEY, response.getHeaderData());
+      return new PickSubchannelArgsImpl(args.getMethodDescriptor(), headers, args.getCallOptions());
+    }
+
+    private PickResult handleError(PickSubchannelArgs args, Status cause) {
+      RequestProcessingStrategy strategy =
+          lbPolicyConfig.getRouteLookupConfig().getRequestProcessingStrategy();
+      switch (strategy) {
+        case SYNC_LOOKUP_CLIENT_SEES_ERROR:
+          return PickResult.withError(cause);
+        case SYNC_LOOKUP_DEFAULT_TARGET_ON_ERROR:
+          return useFallback(args);
+        default:
+          throw new AssertionError("Unknown RequestProcessingStrategy: " + strategy);
+      }
+    }
+
+    private ChildPolicyWrapper fallbackChildPolicyWrapper;
+
+    /** Uses Subchannel connected to default target. */
+    private PickResult useFallback(PickSubchannelArgs args) {
+      String defaultTarget = lbPolicyConfig.getRouteLookupConfig().getDefaultTarget();
+      if (fallbackChildPolicyWrapper == null
+          || !fallbackChildPolicyWrapper.getTarget().equals(defaultTarget)) {
+        try {
+          startFallbackChildPolicy()
+              .await(
+                  lbPolicyConfig.getRouteLookupConfig().getLookupServiceTimeoutInMillis(),
```

Comment:
Why block for fallback policy for `lookupServiceTimeoutInMillis`, which is the timeout value for lookup service requests?

Actually this latch normally is release immediately unless there is some hanging task in the queue of sychronizationConxtext.